### PR TITLE
3233/caworker auth

### PR DESCRIPTION
**Pull request by cornelinux**

### CODE REVIEW

File: privacyidea/lib/tokens/certificatetoken.py [hunk 1]
```diff
@@ -599,7 +605,8 @@ def revoke(self):
 
         # call CAConnector.revoke_cert()
         ca_obj = get_caconnector_object(ca_specifier)
-        revoked = ca_obj.revoke_cert(certificate_pem)
+        revoked = ca_obj.revoke_cert(certificate_pem,
+                                     request_id=ti.get("requestId"))
```
Comment (plettich) [inline privacyidea/lib/tokens/certificatetoken.py:609]: Please use the `REQUEST_ID` constant here

File: tests/test_lib_caconnector.py [hunk 2]
```diff
@@ -468,10 +469,10 @@ def test_01_create_ca_connector(self):
             self.assertEqual(description[key], "string")
 
         # Check, if an error is raised if a required attribute is missing:
-        self.assertRaisesRegexp(CAError, "required argument 'port' is missing.",
-                                MSCAConnector, "billsCA", {MS_ATTR.HOSTNAME: "hans"})
-        self.assertRaisesRegexp(CAError, "required argument 'hostname' is missing.",
-                                MSCAConnector, "billsCA", {MS_ATTR.PORT: "shanghai"})
+        self.assertRaisesRegex(CAError, "required argument 'port' is missing.",
```
Comment (plettich) [inline tests/test_lib_caconnector.py:472]: Please stick with `assertRaisesRegexp()` for now. We can change it when we officially drop Python 2.

File: tests/test_lib_caconnector.py [hunk 3]
```diff
@@ -555,6 +557,54 @@ def test_04_test_pending_request(self):
             r = cacon.get_issued_certificate(request_id)
             self.assertTrue(r.startswith("-----BEGIN CERTIFICATE-----"), r)
 
+    def test_10_fail_ca_missing_certs(self):
```
Comment (plettich) [inline tests/test_lib_caconnector.py:560]: You can add the `@unittest.skipUnless` decorator to the `MSCATestCase` class so you don't have to add it to every function.

File: privacyidea/lib/caconnectors/msca.py [hunk 4]
```diff
@@ -20,52 +20,62 @@
...
 
 from privacyidea.lib.error import CAError
 from privacyidea.lib.caconnectors.baseca import BaseCAConnector, AvailableCAConnectors
 from OpenSSL import crypto
 import logging
+import traceback
 from six.moves import input
 from privacyidea.lib.utils import is_true
 from privacyidea.lib.error import CSRError, CSRPending
+from OpenSSL.crypto import dump_privatekey, load_privatekey
```
Comment (plettich) [inline privacyidea/lib/caconnectors/msca.py:33]: We should use [`pyca/cryptography`](https://cryptography.io/en/latest/) for x509 certificate handling. The `PyOpenSSL` projects states that it should only be used for SSL/TLS handling: https://pypi.org/project/pyOpenSSL/

**Review by plettich — CHANGES_REQUESTED** [review]

File: doc/configuration/caconnectors.rst [hunk 5]
```diff
@@ -164,15 +167,65 @@ The port on which the worker listens.
...
+
+**CA certificate**
+
+This is the location of the file, that contains the CA certificate, that issued the
+CA worker servider certificate. This file is located on the privacyIDEA server in PEM format.
+
+**Client certificate**
+
+This is the file locateion of the certificate that privacyIDEA uses to authenticate against the CA worker.
+It is PEM format.
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/configuration/caconnectors.rst:195]: will fix typo

File: doc/configuration/caconnectors.rst [hunk 6]
```diff
@@ -164,15 +167,65 @@ The port on which the worker listens.
...
+   be used for authentication.
+
+**CA certificate**
+
+This is the location of the file, that contains the CA certificate, that issued the
+CA worker servider certificate. This file is located on the privacyIDEA server in PEM format.
+
+**Client certificate**
+
+This is the file locateion of the certificate that privacyIDEA uses to authenticate against the CA worker.
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/configuration/caconnectors.rst:194]: will fix typo

[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/configuration/caconnectors.rst:194]: done in 74ee87c4

File: doc/configuration/caconnectors.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (cornelinux) [reply doc/configuration/caconnectors.rst:195]: done in 74ee87c4

File: privacyidea/lib/caconnectors/localca.py [hunk 7]
```diff
@@ -441,7 +441,7 @@ def sign_request(self, csr, options=None):
         else:
             filetype = crypto.FILETYPE_PEM
         cert_obj = crypto.load_certificate(filetype, certificate)
-        return cert_obj
+        return 0, cert_obj
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/localca.py:444]: done in ce000bf8

File: privacyidea/lib/caconnectors/msca.py [hunk 8]
```diff
@@ -177,24 +247,42 @@ def sign_request(self, csr, options=None):
...
-                raise CSRError()
+                log.warning("certification request could not be fulfilled! {0!s}".format(reply))
+                raise CSRError(description=reply.dispositionMessage)
+
+    def revoke_cert(self, certificate, request_id=None, reason=None):
+        """
+        Revoke the specified certificate. At this point only the database
+        index.txt is updated.
+
+        :param certificate: The certificate to revoke
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:276]: done in ce000bf8

File: privacyidea/lib/caconnectors/msca.py [hunk 9]
```diff
@@ -177,24 +247,42 @@ def sign_request(self, csr, options=None):
         :type csr: PEM string or SPKAC
         :param options: Additional options like the validity time or the template or spkac=1
         :type options: dict
-        :return: Returns the certificate object if cert was provided instantly
-        :rtype: X509 or None
+        :return: Returns a tuple of requestID and the certificate object if cert was provided instantly
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:252]: You see :-) Added it in the lacalca.py, also.

File: privacyidea/lib/caconnectors/msca.py [hunk 10]
```diff
@@ -119,18 +139,58 @@ def get_config(self, config):
...
+            channel = grpc.insecure_channel('{0!s}:{1!s}'.format(self.hostname, self.port),
+                                            options=(('grpc.enable_http_proxy', int(is_true(self.http_proxy))),))
         try:
-            grpc.channel_ready_future(channel).result(timeout=3)
+            grpc.channel_ready_future(channel).result(timeout=TIMEOUT)
         except grpc.FutureTimeoutError:
             log.warning("Worker seems to be offline. No connection could be established!")
+            log.info(traceback.format_exc())
+        except UnboundLocalError:
+            log.warning("Channel to MS CA worker was not set up successfully.")
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:195]: add traceback debug in df3caf21

File: privacyidea/lib/caconnectors/msca.py [hunk 11]
```diff
@@ -177,24 +247,42 @@ def sign_request(self, csr, options=None):
...
                 raise CSRPending(requestId=reply.requestId)
             else:
-                log.error("certification request could not be fulfilled! {0!s}".format(reply))
-                raise CSRError()
+                log.warning("certification request could not be fulfilled! {0!s}".format(reply))
+                raise CSRError(description=reply.dispositionMessage)
+
+    def revoke_cert(self, certificate, request_id=None, reason=None):
+        """
+        Revoke the specified certificate. At this point only the database
```
[the remark this answers is not in the document]
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py:274]: This is why there is a TODO. See #3316 
I would like to keep it currently this way to also understand, what needs to be implemented.

File: tests/test_lib_caconnector.py (hunk 2 above)
Comment (cornelinux) [reply tests/test_lib_caconnector.py:472]: Oh, I changed this at a lot of places, I think.
Didn't we already drop the test cases for Python2?

File: tests/test_lib_caconnector.py (hunk 3 above)
Comment (cornelinux) [reply tests/test_lib_caconnector.py:560]: Looks to me as if this is not needed for `test_10...`.

File: privacyidea/lib/caconnectors/msca.py (hunk 4 above)
Comment (cornelinux) [reply privacyidea/lib/caconnectors/msca.py]: OK, I think I got it
Done in 85bc05c6

File: tests/test_lib_caconnector.py (hunk 2 above)
Comment (plettich) [reply tests/test_lib_caconnector.py:472]: Nope, just for Python 3.5

**Review by plettich — CHANGES_REQUESTED** [review]

**Pull request merged**